internal/hcsoci/drivers.go

```diff
@@ -0,0 +1,109 @@
+// +build windows
+
+package hcsoci
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/Microsoft/hcsshim/internal/oci"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+	"github.com/pkg/errors"
+)
+
+// getAssignedDeviceUtilityTool gets the path to device-util on the server host
+func getAssignedDeviceUtilityTool(coi *createOptionsInternal) (string, error) {
+	tools, ok := coi.Spec.Annotations[oci.AnnotationAssignedDeviceUtilityTool]
+	if !ok || tools == "" {
+		return "", fmt.Errorf("no driver tools were specified for %s", coi.actualID)
+	}
+	if _, err := os.Stat(tools); err != nil {
+		return "", errors.Wrapf(err, "failed to find device installation tools at %s", tools)
+	}
+	return tools, nil
+}
+
+// getAssignedDeviceKernelDrivers gets any device drivers specified on the spec.
+// Drivers are optional, therefore do not return an error if none are on the spec.
+func getAssignedDeviceKernelDrivers(coi *createOptionsInternal) ([]string, error) {
+	csDrivers, ok := coi.Spec.Annotations[oci.AnnotationAssignedDeviceKernelDrivers]
+	if !ok || csDrivers == "" {
+		return nil, nil
+	}
+	drivers := strings.Split(csDrivers, ",")
+	for _, driver := range drivers {
+		if _, err := os.Stat(driver); err != nil {
+			return nil, errors.Wrapf(err, "failed to find path to drivers at %s", driver)
+		}
+	}
+	return drivers, nil
+}
+
+// setupDeviceUtilTool finds the utility tool's host path, mounts it using vsmb, and
+// returns the UVM path to the tools
+func setupDeviceUtilTool(ctx context.Context, coi *createOptionsInternal, r *Resources) (string, error) {
+	toolHostPath, err := getAssignedDeviceUtilityTool(coi)
+	if err != nil {
+		return "", err
+	}
+	return addVSMBToUVM(ctx, coi.HostingSystem, r, toolHostPath)
+}
+
+// installWindowsDrivers finds specified kernel driver directories, mounts them using vsmb,
+// then installs them in the UVM
+func installWindowsDrivers(ctx context.Context, coi *createOptionsInternal, resources *Resources) error {
```

Comment:
I don't like how tightly coupled this code is to the internals of `hcsoci`, and would like us to try to reduce that coupling and perhaps even move it out to a separate package.

For instance, we are passing in `coi` and `resources` here, but it appears we are only using `coi` to look up an annotation, and to access the UVM object. Likewise, we only use `resources` to add the SCSI resources to the list for clean up later.

I would propose this function instead take:
- The value of the annotation, already extracted from the spec.
- The UVM object.
And return a set of resources to be tracked, rather than adding them to `resources` itself.

Reply:
I like the idea of changing the function args, but I'm not sure how I feel about returning a list of resources. I'd rather resources be added to `resources` immediately after being mounted/added/etc to the UVM.

Reply:
I think a lot of the value in the `ResourceCloser` type is that the code that sets up the resource doesn't need to be coupled with `hcsoci` and the code that tracks the resources to free later. I think the design we should aim for is to have various functions that set up resources and then return a list of the resources so they can be freed later. That function would then be called by code in `hcsoci` which receives the list of resources and adds it to the tracking list so they are cleaned up later.

An example of this is how we manage assigning resources like VSMB/SCSI/etc. to the UVM. The VSMB assignment function returns a resource that is then added to the list by `hcsoci`. If the VSMB code itself was responsible for adding the resource to the tracking list, then all of that code would also have to live in the `hcsoci` package.

Reply:
After talking about this offline, the goal here is to break device related code into a different package and minimize the arguments used to only those necessary. The difficulty of moving devices code out is that the Resources struct lives in the hcsoci pacakge and the hcsoci code that creates wcow and lcow documents call into the devices code. Causing a circular dependency. This can be remedied by moving some more of the code in hcsoci out into logical packages. This will likely be done in a separate PR that this one will rebase on top of to reduce the size of this one.

Reply:
That PR has been merged, this PR has been rebased